### PR TITLE
Bring back netstandard 2.0 support.

### PR DESCRIPTION
**Pull request by RobThree**
This PR brings back Netstandard 2.0 support and, with it, also .Net Framework 4.6+ support. All at the "cost" of a few preprocessor directives with before-and-after .Net 5.0 implementations of some code.

Only one (internal) class was added (`Internals.cs`) which provides record support (`IsExternalInit`) as well as `Range` and `Index` support. The rest is all pretty straightforward.

Everything keeps working as expected; it should not introduce any breaking changes and as an added bonus the older platforms now also get Async support which isn't in 2.x.

* I made the CLI application multi targetted (.Net Framework 4.8 and .Net 5.0) and also async (it still was synchronous).
* I added `Internals.cs` with some shims for records/range/index
* I upgraded the unittest project to .Net 6.0 to get rid of a ".Net 5.0 is no longer supported" warning
* In `NtpClient.cs` I made a few members explicitly private where it wasn't explicitly done so (which is the default, it just is more in line with the rest of the code) and I refactored code a teeny bit to accommodate the conditional code in the preprocessor directives and support async
* In `NtpClock.cs` I refactored the code to work with ticks pre-.Net 5.0 because the `*` operator isn't overloaded pre-.Net 5.0
* In `NtpDateTime.cs` I used ticks because microseconds aren't set by `FromSeconds()` in pre-.Net 5.0 projects
* In `NtpPaclet.cs` I use a different overload of `Enum.IsDefined()` to work in pre-.Net 5.0 projects

### CODE REVIEW

File: GuerrillaNtp/GuerrillaNtp.csproj [hunk 1]
```diff
@@ -1,21 +1,27 @@
...
+		<LangVersion>latest</LangVersion>
+		<EnableNETAnalyzers>true</EnableNETAnalyzers>
+		<AnalysisLevel>latest</AnalysisLevel>
+	</PropertyGroup>
+	<ItemGroup>
+		<None Include="../README.md" Pack="true" PackagePath="/" />
+		<None Include="icon.png" Pack="true" PackagePath="/" />
+	</ItemGroup>
+	<ItemGroup>
+	  <PackageReference Include="System.Memory" Version="4.5.5" />
```
Comment (robertvazan) [inline GuerrillaNtp/GuerrillaNtp.csproj:26]: Shouldn't this be conditional on .NET Standard build?

File: GuerrillaNtp/Internals.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
+﻿#nullable enable
+#if !NET5_0_OR_GREATER
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    // https://github.com/dotnet/runtime/blob/419e949d258ecee4c40a460fb09c66d974229623/src/libraries/System.Private.CoreLib/src/System/Index.cs
```
Comment (robertvazan) [inline GuerrillaNtp/Internals.cs:7]: Impact on licensing?

**Review by robertvazan — COMMENTED** [review]
Looks promising. How is it tested? I see unit tests are targeting .NET 6, so they never run .NET Standard code. <s>Ditto for the CLI</s>.

**Comment by robertvazan** [comment]
Could you bump .NET version in `.github/workflows/build.yml` in your PR, so that we can see the build pass at least for .NET 6?

File: GuerrillaNtp/GuerrillaNtp.csproj (hunk 1 above)
Comment (RobThree) [reply GuerrillaNtp/GuerrillaNtp.csproj:26]: > Looks promising. How is it tested? I see unit tests are targeting .NET 6, so they never run .NET Standard code. Ditto for the CLI.

You can run the Cli the .Net 4.8 or .Net 5.0 directory in the bin/debug directories.

> Shouldn't this be conditional on .NET Standard build?

No. `System.Memory` adds support for some older platforms and in newer platforms it's just a NOP.

File: GuerrillaNtp/Internals.cs (hunk 2 above)
Comment (RobThree) [reply GuerrillaNtp/Internals.cs:7]: > Impact on licensing?

It's not a copy/paste but a derivative. IANAL but I've seen it used in a lot more projects. I'm quite sure it should be OK, if not, there are 3rd party NuGet packages that do the same.

**Comment by robertvazan** [comment]
Even though the PR is comprehensive and high quality, there is still work left to do on my side. Automated build will have to run for multiple .NET versions. This could take some time as I am quite busy. Release script will need update too. I am starting to wonder whether it wouldn't be easier to fork and submit to NuGet under name that indicates it is the same thing with .NET Standard compatibility.

**Comment by RobThree** [comment]
> I am starting to wonder whether it wouldn't be easier to fork and submit to NuGet under name that indicates it is the same thing with .NET Standard compatibility.

I think that would be a shame. It's exactly the same library as it is in the current release, just with added (back) support for older platforms and as an added bonus the Async support that older platforms didn't have. This shouldn't have to be more than a minor version bump IMHO. Submitting it to NuGet under a new name would cause unnecessary confusion and fragmentation if you ask me.

File: GuerrillaNtp/Internals.cs (hunk 2 above)
Comment (robertvazan) [reply GuerrillaNtp/Internals.cs:7]: Way too similar to be original work. It is not necessary to have this Apache-licensed though. Apache license has NOTICE file, which can mention license for incorporated code and whether/how it was modified.

Comment (RobThree) [reply GuerrillaNtp/Internals.cs:7]: > Way too similar

Well, it kind of has to be pretty similar for it to work without breaking the API's implemented in this class 😉 Having said that, again, IANAL but all I can say is that > I < wouldn't worry too much about it 😜

**Comment by robertvazan** [comment]
> I think that would be a shame. It's exactly the same library as it is in the current release, just with added (back) support for older platforms and as an added bonus the Async support that older platforms didn't have. This shouldn't have to be more than a minor version bump IMHO. Submitting it to NuGet under a new name would cause unnecessary confusion and fragmentation if you ask me.

I understand that. It's just that I might have trouble finding time for this in the next few weeks if there's any non-trivial work left to do.

**Comment by RobThree** [comment]
> if there's any non-trivial work left to do.

I don't know if there's any non-trivial work to do, I can't speak for that or for you for that matter. I don't know how this is tested other than the included unittest project (which is pretty bare). All I can say is that I've tested it under .Net 4.8, .Net 5.0 and .Net 6.0 and all worked fine without any breaking changes. So I'll leave it up to you. If there's anything I can do to help, I would gladly (try to) do so. Just ask.

**Comment by robertvazan** [comment]
Note to self regarding work left to do:

- multi-release CI build
- multi-release unit tests
- multi-release nuget push
- mention imported code in NOTICE
- update project setup scripts to match manual csproj edits

**Comment by robertvazan** [comment]
@RobThree You could definitely help with the multi-release unit tests. Unit tests are not as bare as they seem. They query actual NTP server, which does request validation, and then the response goes through GuerrillaNtp's validation code. Brief mention of imported code in NOTICE would also help, just to satisfy licensing requirements. The rest is up to me.

I will try to find time for this in the near future, but I cannot promise any timeline. Once merged, I will make a release.

**Comment by RobThree** [comment]
> Note to self regarding work left to do:

> multi-release CI build
> multi-release nuget push

You don't need to do that? The `.csproj` already has 2 framework targets. Those are just put in the single NuGet package and uploaded as one. You don't need to push different packages to NuGet for each platform.

> Unit tests are not as bare as they seem. They query actual NTP server, which does request validation, and then the response goes through GuerrillaNtp's validation code.

I noticed, but to be honest: I don't think a unittest should call into external resources anyway. These should be mocked or whatever.

> Brief mention of imported code in NOTICE would also help, just to satisfy licensing requirements.

~I'll do that.~ [Done](https://github.com/robertvazan/guerrillantp/pull/12/commits/6b56d89dfc3240db43796ab21ec9da9aa407ee81)

**Comment by robertvazan** [comment]
> You don't need to do that? The `.csproj` already has 2 framework targets. Those are just put in the single NuGet package and uploaded as one. You don't need to push different packages to NuGet for each platform.

But the release GitHub Action must first install the relevant .NET versions, which is why it asks for a list of .NET versions I think.

**Comment by robertvazan** [comment]
@RobThree There's test failure under .NET 4.8. Could you please take a look at it?

**Comment by RobThree** [comment]
> @RobThree There's test failure under .NET 4.8. Could you please take a look at it?

Currently looking into it - also how to run unittests for .Net 4.8 and 6.0 (this works out of the box for MSTest and I see both frameworks in my Test Explorer, but NUnit only runs one...?)

**Comment by robertvazan** [comment]
> Currently looking into it - also how to run unittests for .Net 4.8 and 6.0 (this works out of the box for MSTest and I see both frameworks in my Test Explorer, but NUnit only runs one...?)

No idea. But CI build ran both automatically without being configured to do so.

**Comment by RobThree** [comment]
> > Currently looking into it - also how to run unittests for .Net 4.8 and 6.0 (this works out of the box for MSTest and I see both frameworks in my Test Explorer, but NUnit only runs one...?)
> 
> No idea. But CI build ran both automatically without being configured to do so.

Yeah, it appears my Visual Studio is borked or something; which may explain why I didn't notice the (weird, by the way) exception in .Net 4.8. But I'm sure I'll be able to fix it.

Edit: I think it's related to [this](https://github.com/mono/mono/issues/8967#issue-327967104).

**Comment by RobThree** [comment]
I have no idea what's going on. I made only [this tiny change](https://github.com/RobThree/guerrillantp/commit/b4a8a914cf6093cfddf12c55aa4df72ea35d5559) and somehow the test has now been running for (currently) [over 11 minutes](https://github.com/robertvazan/guerrillantp/actions/runs/3533910953/jobs/5930200825)... Not sure if this is just a fluke / something borked... But it seems stuck...

Anyway, I'm going to have dinner and a little family time and will look into this again later tonight. However, I believe this is only in the build environment (assuming it's using Mono to run .Net Framework projects) and so only when ConnectAsync is invoked in .Net Framework projects on non-Windows platform. But I can't stand it being like this, so I'm going to look into this further.

**Comment by RobThree** [comment]
Ok, I'm an incredible dumbass. For one, I didn't notice that besides the `DnsEndpoint` an `IPEndpoint` was used. And once I noticed that I didn't notice that it wasn't the `ConnectAsync()` method that ("kept") throwing, it was the `Connect()` method - which, naturally, required the same fix.

**Comment by RobThree** [comment]
Ok, it [builds and tests fine](https://github.com/RobThree/guerrillantp/actions) with .Net 5.0 / .Net 6.0 now. However, putting `<TargetFrameworks>net4.8;net6.0</TargetFrameworks>` in the `GuerrillaNtp.Tests.csproj` seems to get the testproject stuck. Everything works fine, here, on my end (on a Windows machine) under both .Net 4.8, .Net 5.0 and up.

I figure the tests getting stuck for .Net 4.8 it has something to to with the `build.yml` where maybe 2 .Net frameworks need to be specified? Or maybe two separate build actions? But I'm not even entirely sure if GitHub even _can_ run tests for .Net Framework. I'm afraid I'll have to rely on you for that.

My apologies for the confusion and the many commits; in the end nothing much changed except the separate branch for the `DnsEndpoint` in the code.

**Comment by RobThree** [comment]
Ok, I see now why you said fixing the build scripts might take some effort 😅 

I've been tinkering a bit with the .Net versions (see above commits). One reason why I don't like Actions is because they're so fiddly to set up and it's hard to test them without committing.

Anyway... 

* The library itself is now configured as it should be (in my opinion). I have added a `<CheckEolTargetFramework>false</CheckEolTargetFramework>` to stop VS complaining about .Net 5 being EOL. It also compiles and works for .Net 4.8 / netstandard 2.0 but I can't figure out why the Github action gets stuck
* The CLI and Unittest project are currently configured for .Net 5.0 but should ideally be .Net 6 or even .Net 7. This, however, requires two ,Net versions to be installed for everything to compile.

One of the easier solutions may be stop explicitly building for .Net 5. We could have .Net 5 use the netstandard2.0 code and change the preprocessor directives to .Net 6 and build for netstandard2.0 and .Net 6.0. The library is, functionally, the same for pre/post .Net 5 code so that shouldn't be a problem.

The other solution would be that someone else much smarter than me tries to fix the github actions.

Either way; besides soms build-trouble I think the code is in good shape and the library is in it's final form (as far as bringing back pre-.Net 5.0 support and still being compatible).

When this is fixed, before releasing, please let me know. I have (on another branch) one other tiny change: [`nullable` support](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-reference-types) see [here](https://github.com/RobThree/guerrillantp/commit/e258861e8de2de3a79995c4771845dad99ab4156#diff-17bb3351010cbde6323aa0d9edf3d49dd8fccc54356c386fdf09e795de423be1).

There's no hurry, but I would *really* appreciate my work being merged and released. Thanks, again, for a great library!

**Comment by robertvazan** [comment]
@RobThree Thank you. I will merge, clean up, and release as soon as I have a bit of time.

**Comment by RobThree** [comment]
Thank you very much. Much appreciated. Take your time.

**Comment by robertvazan** [comment]
Released in version 3.1.0. Let me know if there are any issues.

**Comment by robertvazan** [comment]
@RobThree I am cleaning up the project and I am wondering whether there is some way to run tests against the .NET Standard code, either locally or in CI action. How did you test your changes?

**Comment by RobThree** [comment]
Hmm, I was convinced the Test project targetted both .Net 4.&lt;something&gt; and netstandard 2.0. But [apparently not](https://github.com/robertvazan/guerrillantp/blob/f816dae2e9bd90e5aa84f8c7f60f6489c406b426/GuerrillaNtp.Tests/GuerrillaNtp.Tests.csproj#L4). I'm not sure if it simply wasn't committed or... Let me see.

**Comment by RobThree** [comment]
Hmm, when I make the following change:

image

I can run the tests individually from VS2022 and everything works.

image

When I run them all at once some unittest simply hang / deadlock. However, running the tests from a commandline everything is ok again:

image

I have no idea what's going on. I'm not sure how I tested back then, but I'm 99% sure I did test netstandard and net framework. I'm also not sure if this is a problem with my VS2022 or something else is going on. Unfortunately my memory is a little hazy on  what happened in November last year 😬 

Edit:

> When I run them all at once some unittest simply hang / deadlock.

And now they don't (anymore) 🤪  All tests run fine now...

image

**Comment by robertvazan** [comment]
@RobThree I removed .NET 4.8 target from test project, because there is no way to run it on Linux (and I am developing on Linux). Is there another way to test .NET Standard code?

GitHub runner that I am currently using (ubuntu-latest) cannot run it either. Perhaps Windows runner could, but I would have to develop and debug another workflow. I am not sure about impact of having unsupported framework in the list on local and CI builds. Do you think there is an easier way?

**Comment by RobThree** [comment]
Wait. I've been confusing myself here. This was never about .Net framework vs .Net (core). This was about netstandard. Let me look into this tonight after work or maybe in a few days; I'm currently trying to focus on my (work) project and am confusing things here. Sorry. I'll get back to this ASAP.

**Comment by robertvazan** [comment]
Take your time. If you cannot find a solution, I will probably go for the new workflow sometime later.

**Pull request merged**